### PR TITLE
fix(eks): circular dependencies when security groups from other stacks are used

### LINKED ISSUE #9536: [aws-eks] cannot define eks cluster and network in different stacks
in v1.57.0 stack reference doesn't work, everything is fine in v1.56.0

### Reproduction Steps

```
network_sg_stack = NetworkSgStack(
    scope=app,
    id='sg-stack-' + environment,
    env=env,
    environment=environment,
    vpc_id=app.node.try_get_context(target_env)['vpc_id'],
)

eks_stack = EksStack(
    scope=app,
    id='eks-stack-' + environment,
    env=env,
    environment=environment,
    vpc=network_sg_stack.vpc,
    subnet_a_eks=network_sg_stack.subnet_c_eks,
    subnet_b_eks=network_sg_stack.subnet_b_eks,
    subnet_c_eks=network_sg_stack.subnet_c_eks,
    cluster_version=aws_eks.KubernetesVersion.V1_16
    tags=tags,
    eks_control_plane_role=iam_stack.eks_control_plane_role,
    eks_control_plane_sg=network_sg_stack.eks_control_plane_sg,
)
```
-->

### What actually happened?
```
jsii.errors.JavaScriptError: 
  Error: 'sg-stack-develop-cdk2' depends on 'edp-eks-stack-develop-cdk2' (sg-stack-develop-cdk2 -> eks-stack-develop-cdk2/cluster/KubectlProviderSecurityGroup/Resource.GroupId). Adding this dependency (stack-develop-cdk2 -> sg-stack-develop-cdk2/eks-control-plane-sg-develop-cdk2/Resource.GroupId) would create a cyclic reference.
      at Stack._addAssemblyDependency (/tmp/jsii-kernel-mpylii/node_modules/@aws-cdk/core/lib/stack.js:415:19)
      at Object.addDependency (/tmp/jsii-kernel-mpylii/node_modules/@aws-cdk/core/lib/deps.js:47:24)
      at Stack.addDependency (/tmp/jsii-kernel-mpylii/node_modules/@aws-cdk/core/lib/stack.js:194:16)
      at resolveValue (/tmp/jsii-kernel-mpylii/node_modules/@aws-cdk/core/lib/private/refs.js:85:14)
      at Object.resolveReferences (/tmp/jsii-kernel-mpylii/node_modules/@aws-cdk/core/lib/private/refs.js:27:30)
      at Object.prepareApp (/tmp/jsii-kernel-mpylii/node_modules/@aws-cdk/core/lib/private/prepare-app.js:34:16)
      at Object.synthesize (/tmp/jsii-kernel-mpylii/node_modules/@aws-cdk/core/lib/private/synthesis.js:16:19)
      at App.synth (/tmp/jsii-kernel-mpylii/node_modules/@aws-cdk/core/lib/stage.js:78:41)
      at /home/alexey/.local/lib/python3.7/site-packages/jsii/_embedded/jsii/jsii-runtime.js:7746:51
      at Kernel._wrapSandboxCode (/home/alexey/.local/lib/python3.7/site-packages/jsii/_embedded/jsii/jsii-runtime.js:8398:19)
      at /home/alexey/.local/lib/python3.7/site-packages/jsii/_embedded/jsii/jsii-runtime.js:7746:25
      at Kernel._ensureSync (/home/alexey/.local/lib/python3.7/site-packages/jsii/_embedded/jsii/jsii-runtime.js:8371:20)
      at Kernel.invoke (/home/alexey/.local/lib/python3.7/site-packages/jsii/_embedded/jsii/jsii-runtime.js:7745:26)
      at KernelHost.processRequest (/home/alexey/.local/lib/python3.7/site-packages/jsii/_embedded/jsii/jsii-runtime.js:7446:28)
      at KernelHost.run (/home/alexey/.local/lib/python3.7/site-packages/jsii/_embedded/jsii/jsii-runtime.js:7384:14)
      at Immediate._onImmediate (/home/alexey/.local/lib/python3.7/site-packages/jsii/_embedded/jsii/jsii-runtime.js:7387:37)
      at processImmediate (internal/timers.js:456:21)

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "app.py", line 105, in <module>
    app.synth()
  File "/home/alexey/.local/lib/python3.7/site-packages/aws_cdk/core/__init__.py", line 11426, in synth
    return jsii.invoke(self, "synth", [options])
  File "/home/alexey/.local/lib/python3.7/site-packages/jsii/_kernel/__init__.py", line 113, in wrapped
    return _recursize_dereference(kernel, fn(kernel, *args, **kwargs))
  File "/home/alexey/.local/lib/python3.7/site-packages/jsii/_kernel/__init__.py", line 296, in invoke
    return _callback_till_result(self, response, InvokeResponse)
  File "/home/alexey/.local/lib/python3.7/site-packages/jsii/_kernel/__init__.py", line 181, in _callback_till_result
    response = kernel.sync_complete(response.cbid, None, result, response_type)
  File "/home/alexey/.local/lib/python3.7/site-packages/jsii/_kernel/__init__.py", line 339, in sync_complete
    response_type=response_type
  File "/home/alexey/.local/lib/python3.7/site-packages/jsii/_kernel/providers/process.py", line 369, in sync_complete
    resp = self._process.send(_CompleteRequest(complete=request), response_type)
  File "/home/alexey/.local/lib/python3.7/site-packages/jsii/_kernel/providers/process.py", line 318, in send
    raise JSIIError(resp.error) from JavaScriptError(resp.stack)
jsii.errors.JSIIError: 'sg-stack-develop-cdk2' depends on 'eks-stack-develop-cdk2' (sg-stack-develop-cdk2 -> eks-stack-develop-cdk2/cluster/KubectlProviderSecurityGroup/Resource.GroupId). Adding this dependency (eks-stack-develop-cdk2 -> sg-stack-develop-cdk2/eks-control-plane-sg-develop-cdk2/Resource.GroupId) would create a cyclic reference.
Subprocess exited with error 1
```

### Environment

  - **CLI Version      *1.57.0*
  - **Framework Version:*1.57.0*
  - **Node.js Version:*v12.18.1*
  - **OS               :ubuntu 18.04
  - **Language (Version):*Python 3.7.5*

### Other

--- 

This is :bug: Bug Report

### LINKED ISSUE #9754: [aws-eks] eks cluster connections refers to wrong security group
After creating an EKS cluster with a default Managed Node Group via the `aws-eks` module's `Cluster()` constructor, I noticed that if I tried to allow ingress to the nodes and pods on the Node Group via `cluster.connections` that it failed to work. This is because the security group associated with `cluster.connections` is the _secondary_ security group that's being assigned to the cluster via its resourcesVpcConfig, instead of the _primary_ cluster security group that's returned by the EKS CreateCluster API. 

EKS Managed Node Groups and the pods that run on them are associated with the cluster security group that's returned by the EKS cluster API, **not** any additional security groups that a customer may configure.

### Reproduction Steps

```
const cluster = new eks.Cluster(this, clusterName, {
    clusterName,
    version: eks.KubernetesVersion.V1_17,
    mastersRole: accessRole,
    vpc,
    vpcSubnets: [
      { subnetType: ec2.SubnetType.PRIVATE },
    ],
    defaultCapacity: 2
  }
));

cluster.connections.allowFrom(externalClient.connections, ec2.Port.allTcp());
```
### What did you expect to happen?

I expected an ingress rule to be associated with the cluster security group associated with my cluster that allowed ingress from all TCP ports from my external client's security group.

### What actually happened?

An ingress rule was added to the _additional security group_ associated with my cluster that allowed ingress from all TCP ports from my external client's security group. Unfortunately this has no effect on the pods running in the cluster.

### Environment

  - **CLI Version      :** 1.56.0
  - **Framework Version:** 1.56.0
  - **Node.js Version:**  12.18.2
  - **OS               :** MacOS Catalina
  - **Language (Version):** TypeScript (3.8.3)

### Other

The following code provides an interim workaround:

```
  new ec2.Connections({
    securityGroups: [
      ec2.SecurityGroup.fromSecurityGroupId(this, 
           `${clusterName}ClusterSecurityGroup`, cluster.clusterSecurityGroupId)
      ]
  }).allowFrom(externalClient.connections, ec2.Port.allTcp());
```

--- 

This is :bug: Bug Report

### LINKED ISSUE #10020: [aws-eks] Multiple object reference to Nested Stack causes Circular Dependency.
## :question: General Issue

The `Cluster` interface creates circular dependency errors when multiple references to differing `NestedStack` child-objects.  I'm not sure if this is an implementation error with `Cluster` or an inherent design flaw with `NestedStack`.  It seems like a DAG would solve for this issue.

## Tool Spec

* `node@12`
* `cdk@1.6`
* `language=js`

## Error

``` bash
Circular dependency between resources: [ControlPlaneEKSCreationRole58D1869A, ControlPlaneEKSKubectlReadyBarrier1C298D16, ControlPlaneEKSKubectlProviderSecurityGroup68FF96CB, ControlPlaneEKSCreationRoleDefaultPolicy5BFFD31E, NetworkNestedStackNetworkNestedStackResource0124E108, ControlPlaneEKSAwsAuthmanifestD1CB63E6, ControlPlaneEKSC6E4919A]
```

## Code Snippet

``` javascript
const { App, Construct, Stack } = require('@aws-cdk/core');
const { Cluster } = require('@aws-cdk/aws-eks');
const { SecurityGroup, Vpc } = require('@aws-cdk/aws-ec2');

class MyApp extends App {
    constructor(scope, id, props) {
        super(scope, id, props);

        new MyStack(this, 'MyUnderlay');
    }
}

class MyStack extends Stack {
    constructor(scope, id, props) {
        super(scope, id, props);

        const { vpc, eksControlPlaneSecurityGroup } = new Network(this, 'MyNetwork');
        new ControlPlane(this, 'MyControlPlane', { vpc, eksControlPlaneSecurityGroup });
    }
}

class Network extends NestedStack {
    constructor(scope, id, props) {
        super(scope, id, props);

        const vpc = new Vpc(this, 'Vpc', {
            cidr: '20.1.0.0/16',
            enableDnsHostnames: true,
            enableDnsSupport: true,
            maxAzs: 99
        });

        const eksControlPlaneSecurityGroup = new SecurityGroup(this, 'EKSControlPlane', {
            securityGroupName: 'MyControlePlaneSG',
            vpc,
            allowAllOutbound: false
        });

        Object.assign(this, { vpc, eksControlPlaneSecurityGroup });
    }
}

class ControlPlane extends Construct {
    constructor(scope, id, props) {
        super(scope, id, props);

        const { vpc, eksControlPlaneSecurityGroup } = props;

        new Cluster(this, 'MyEKS', {
            clusterName: 'fake-cluster',
            version: KubernetesVersion.V1_17,
            vpc,
            securityGroup: eksControlPlaneSecurityGroup,  // removing this resolves the circular dep error observed
            defaultCapacity: 0
        });
    }
}
```

### PR DESCRIPTION
**Pull request by iliapolo**
This PR resolves a few problem related to circular dependencies when the cluster is configured with security groups from other stacks.

## Problem

Currently, the cluster creates a dedicated security group for the `KubectlProvider` and configures the main control plane security group to allow connections from it:

https://github.com/aws/aws-cdk/blob/f5c5d060ffee19ac3daa47cbf6df8d3563133433/packages/%40aws-cdk/aws-eks/lib/cluster.ts#L918-L924

This eventually creates ingress rules in the stack of control plane security group. These ingress rules refer to the `KubectlProvider` creates in the cluster stack, hence, a dependency if formed between the control plane security group stack, and the cluster stack.

`Control Plane Security Group Stack => Cluster Stack` (Control plane sg needs to allow connections from kubectl sg)

On the other hand, the cluster stack naturally depends on the control plane security group stack because the cluster needs the security group when its created. And so:

`Cluster Stack => Control Plane Security Group Stack` (Cluster needs the security group)
  
When these two stacks are different, a circular dependency is created.

## Solution

Do not create a security group for the `KubectlProvider` at all, and simply re-use the EKS managed security group. This security group is already configured to allow connections to the control plane, so if we simply attach it to the `KubectlProvider`, everything should work.

This avoids the creation of the first dependency direction.

------------------------------------------

In addition, we realized that the [`securityGroup`](https://github.com/aws/aws-cdk/blob/master/packages/@aws-cdk/aws-eks/lib/cluster.ts#L285) property is actually not needed anymore since EKS will create a security group for us. 

> https://docs.aws.amazon.com/eks/latest/userguide/sec-group-reqs.html

This is in-fact the same SG we mentioned, and can be used both by the `KubectlProvider` and to configure `cluster.connections`.

The `securityGroup` property was renamed to `additionalSecurityGroups` (plural) and is passed to the cluster creation call, but do not participate in the `cluster.connections`. 

Fixes #9754 
Fixes #10020 
Fixes #9536

BREAKING CHANGE: cluster security group configuration has changed and will cause cluster replacement.

* **eks**: `securityGroup` property was deleted. Use `securityGroups` instead.
* **eks**: `securityGroups` property was added to allow configuring multiple additional security groups on the cluster.

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: packages/@aws-cdk/aws-eks-next/lib/cluster.ts [hunk 1]
```diff
@@ -922,6 +905,15 @@ export class Cluster extends ClusterBase {
     this.clusterSecurityGroupId = resource.attrClusterSecurityGroupId;
     this.clusterEncryptionConfigKeyArn = resource.attrEncryptionConfigKeyArn;
 
+    this.kubectlSecurityGroup = ec2.SecurityGroup.fromSecurityGroupId(this, 'ClusterSecurityGroup', this.clusterSecurityGroupId);
```
Comment (eladb) [inline packages/@aws-cdk/aws-eks-next/lib/cluster.ts:908]: Should we also add `this.clusterSecurityGroup`?

File: packages/@aws-cdk/aws-eks-next/lib/cluster.ts [hunk 2]
```diff
@@ -922,6 +905,15 @@ export class Cluster extends ClusterBase {
...
+
+    this.connections = new ec2.Connections({
+      securityGroups: [this.kubectlSecurityGroup],
```
Comment (eladb) [inline packages/@aws-cdk/aws-eks-next/lib/cluster.ts:911]: It's a bit confusing that you are reusing `kubctlSecurityGroup` here. This should be `this.clusterSecurityGroup` (they could both be the same value but just from a readability perspective it's not clear)

File: packages/@aws-cdk/aws-eks-next/lib/cluster.ts [hunk 3]
```diff
@@ -922,6 +905,15 @@ export class Cluster extends ClusterBase {
...
+      defaultPort: ec2.Port.tcp(443), // Control Plane has an HTTPS API
+    });
+
+    props.securityGroups?.forEach(sg => this.connections.addSecurityGroup(sg));
```
Comment (eladb) [inline packages/@aws-cdk/aws-eks-next/lib/cluster.ts:915]: add a comment explaining why you are doing this

File: packages/@aws-cdk/aws-eks-next/test/test.cluster.ts [hunk 4]
```diff
@@ -19,6 +19,116 @@ const CLUSTER_VERSION = eks.KubernetesVersion.V1_16;
...
+            { Ref: 'VPCPrivateSubnet2SubnetCFCDAA7A' },
+          ],
+        },
+      },
+    }));
+
+    test.done();
+  },
+
+  'additional securityg groups are added to connections on owned cluster'(test: Test) {
```
Comment (eladb) [inline packages/@aws-cdk/aws-eks-next/test/test.cluster.ts:56]:
```suggestion
  'additional security groups are added to connections on owned cluster'(test: Test) {
```

**Review by eladb — APPROVED** [review]
Looks great. Minor comments (added "do-not-merge").

**Comment by iliapolo** [comment]
@eladb Did you see this? https://github.com/aws/aws-cdk/pull/10339

**Comment by iliapolo** [comment]
Closing since this was implemented on master without the need to force cluster replacement.

https://github.com/aws/aws-cdk/pull/10339

**Pull request closed without merging**